src/liballoc/str.rs

```diff
@@ -87,52 +87,102 @@ impl<S: Borrow<str>> SliceConcatExt<str> for [S] {
     type Output = String;
 
     fn concat(&self) -> String {
-        if self.is_empty() {
-            return String::new();
-        }
-
-        // `len` calculation may overflow but push_str will check boundaries
-        let len = self.iter().map(|s| s.borrow().len()).sum();
-        let mut result = String::with_capacity(len);
-
-        for s in self {
-            result.push_str(s.borrow())
-        }
-
-        result
+        self.join("")
     }
 
     fn join(&self, sep: &str) -> String {
-        if self.is_empty() {
-            return String::new();
+        unsafe {
+            String::from_utf8_unchecked( join_generic_copy(self, sep.as_bytes()) )
         }
+    }
+
+    fn connect(&self, sep: &str) -> String {
+        self.join(sep)
+    }
+}
 
-        // concat is faster
-        if sep.is_empty() {
-            return self.concat();
+macro_rules! spezialize_for_lengths {
+    ($separator:expr, $target:expr, $iter:expr; $($num:expr),*) => {
+        let mut target = $target;
+        let iter = $iter;
+        let sep_bytes = $separator;
+        match $separator.len() {
+            $(
+                // loops with hardcoded sizes run much faster
+                // specialize the cases with small separator lengths
+                $num => {
+                    for s in iter {
+                        copy_slice_and_advance!(target, sep_bytes);
+                        copy_slice_and_advance!(target, s.borrow().as_ref());
+                    }
+                },
+            )*
+            _ => {
+                // arbitrary non-zero size fallback
+                for s in iter {
+                    copy_slice_and_advance!(target, sep_bytes);
+                    copy_slice_and_advance!(target, s.borrow().as_ref());
+                }
+            }
         }
+    };
+}
 
-        // this is wrong without the guarantee that `self` is non-empty
-        // `len` calculation may overflow but push_str but will check boundaries
-        let len = sep.len() * (self.len() - 1) +
-                  self.iter().map(|s| s.borrow().len()).sum::<usize>();
-        let mut result = String::with_capacity(len);
-        let mut first = true;
+macro_rules! copy_slice_and_advance {
+    ($target:expr, $bytes:expr) => {
+        let len = $bytes.len();
+        $target.get_unchecked_mut(..len)
+            .copy_from_slice($bytes);
+        $target = {$target}.get_unchecked_mut(len..);
```

Comment:
Can this two `get_unchecked_mut` be replaced with `split_at` perhaps to avoid calling this twice?

Reply:
It seems so. I moved away from `split_at` fairly early in my optimizations when I still used the regular bench suite to gauge speed. I can't see any meaningful improvement now so I might as well change that back.

The benchmarks are incredibly variable and regularly show spurious variations on the order of ~5% and even up to 20% (!). The variations are strangely somewhat persisting across different benchmarks in the same run in that it's almost like a constant factor for all benches. I know they are spurious because they affect the old join just as much as the new.
The variability is what made me switch to the `Criterion` bench framework but that hasn't helped at all.